RecyclerViewKotlin/app/src/main/java/com/example/recyclersample/flowerList/FlowersAdapter.kt

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.recyclersample.flowerList
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import com.example.recyclersample.R
+
+private const val HEADER_TYPE = 0
+private const val FLOWER_TYPE = 1
+
+class FlowersAdapter(private val onClick: (DataItem.FlowerItem) -> Unit) :
+    ListAdapter<DataItem, RecyclerView.ViewHolder>(
+        FlowerDiffCallback()
+    ) {
+
+    class FlowerViewHolder(itemView: View, val onClick: (DataItem.FlowerItem) -> Unit) :
+        RecyclerView.ViewHolder(itemView) {
+        private val flowerTextView: TextView = itemView.findViewById(R.id.flower_text)
+        private val flowerImageView: ImageView = itemView.findViewById(R.id.flower_image)
+        private var currentDataItem: DataItem.FlowerItem? = null
+
+        init {
+            itemView.setOnClickListener {
+                currentDataItem?.let {
+                    onClick(it)
+                }
+            }
+        }
+
+        fun bind(dataItem: DataItem.FlowerItem) {
+            val flower = dataItem.flower
+            currentDataItem = dataItem
+
+            flowerTextView.text = flower.name
+            if (flower.image != null) {
+                flowerImageView.setImageResource(flower.image)
+            } else {
+                flowerImageView.setImageResource(R.drawable.rose)
+            }
+        }
+    }
+
+    class HeaderViewHolder(view: View) : RecyclerView.ViewHolder(view)
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+        return when(viewType){
+            HEADER_TYPE -> {
+                val view = LayoutInflater.from(parent.context)
+                    .inflate(R.layout.header_item, parent, false)
+                HeaderViewHolder(view)
+            }
+            FLOWER_TYPE -> {
+                val view = LayoutInflater.from(parent.context)
+                    .inflate(R.layout.flower_item, parent, false)
+                FlowerViewHolder(view, onClick)
+            }
+            else -> { throw Exception("Invalid viewType") }
+        }
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        if (holder is FlowerViewHolder) {
+            val flower = getItem(position) as DataItem.FlowerItem
+            holder.bind(flower)
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return when (getItem(position)) {
+            is DataItem.HeaderItem -> HEADER_TYPE
+            is DataItem.FlowerItem -> FLOWER_TYPE
+        }
+    }
+}
+
+private class FlowerDiffCallback : DiffUtil.ItemCallback<DataItem>() {
```

Comment:
This could be an `object`

Reply:
looks like our other samples using DiffUtil uses it as a class

Reply:
Sure... but this should still be an `object`. The other samples should probably be updated.

Reply:
updated!